src/IB/IBRedundantInitializer.cpp

```diff
@@ -1370,8 +1414,26 @@ IBRedundantInitializer::getPatchVerticesAtLevel(std::vector<std::pair<int, int>
             std::pair<int, int> point_index(j, k);
             const Point& X =
                 getShiftedVertexPosn(point_index, vertex_level_number, domain_x_lower, domain_x_upper, periodic_shift);
-            const CellIndex<NDIM> idx = IndexUtilities::getCellIndex(X, grid_geom, ratio);
-            if (patch_box.contains(idx)) patch_vertices.push_back(point_index);
+            // For consistency, if a point is located on the top of a cell it is
+            // assigned to the upper, not lower cell. This is problematic at
+            // physical boundaries since points which really should be assigned
+            // to a cell inside the physical domain are not. Work around this by
+            // decrementing the correct component index by one in that case.
+            CellIndex<NDIM> idx = IndexUtilities::getCellIndex(X, grid_geom, ratio);
+            if (patch_on_upper_physical_boundary)
+            {
+                for (int d = 0; d < NDIM; ++d)
+                {
+                    if (X[d] == domain_x_upper[d])
+                    {
+                        idx(d) -= 1;
+                    }
+                }
+            }
+            if (patch_box.contains(idx))
+            {
+                patch_vertices.push_back(point_index);
+            }
```

Comment:
We now permit points that are on the upper part of the domain - here's the logic for that case. Two of the new tests verify that this works.